gpjax/parameters.py

```diff
@@ -144,48 +139,38 @@ def recursive_bijectors(ps, bs) -> tp.Tuple[tp.Dict, tp.Dict]:
     return recursive_bijectors(params, bijectors)
 
 
-def build_transforms(params: tp.Dict) -> tp.Tuple[tp.Dict, tp.Dict]:
-    """Using the bijector that is associated with each parameter, construct a pair of functions from the bijector that allow the parameter to be constrained and unconstrained.
+def constrain(params: tp.Dict, bijectors: tp.Dict) -> tp.Dict:
+    """Transform the parameters to the constrained space for corresponding bijectors.
 
     Args:
-        params (tp.Dict): The parameter set for which transformations should be derived from.
+        params (tp.Dict): The parameters that are to be transformed.
+        transform_map (tp.Dict): The corresponding dictionary of transforms that should be applied to the parameter set.
+        foward (bool): Whether the parameters should be constrained (foward=True) or unconstrained (foward=False).
 
     Returns:
-        tp.Tuple[tp.Dict, tp.Dict]: A pair of dictionaries. The first dictionary maps each parameter to a function that constrains the parameter. The second dictionary maps each parameter to a function that unconstrains the parameter.
+        tp.Dict: A transformed parameter set. The dictionary is equal in structure to the input params dictionary.
     """
 
-    def forward(bijector):
-        return bijector.forward
+    map = lambda param, trans: trans.forward(param)
 
-    def inverse(bijector):
-        return bijector.inverse
+    return jax.tree_util.tree_map(map, params, bijectors)
 
-    bijectors = build_bijectors(params)
-
-    constrainers = jax.tree_util.tree_map(lambda _: forward, deepcopy(params))
-    unconstrainers = jax.tree_util.tree_map(lambda _: inverse, deepcopy(params))
-
-    constrainers = jax.tree_util.tree_map(lambda f, b: f(b), constrainers, bijectors)
-    unconstrainers = jax.tree_util.tree_map(
-        lambda f, b: f(b), unconstrainers, bijectors
-    )
 
-    return constrainers, unconstrainers
-
-
-def transform(params: tp.Dict, transform_map: tp.Dict) -> tp.Dict:
-    """Transform the parameters according to the constraining or unconstraining function dictionary.
+def unconstrain(params: tp.Dict, bijectors: tp.Dict) -> tp.Dict:
```

Comment:
Should these be a function of `params` and `bijectors`, or simply `ParameterState`?

Reply:
That's a nice suggestion.

Reply:
We could have a utility function call `constrain_state(parameter_state: ParameterState)` and `uncontrain_state(parameter_state: ParameterState)` that have the form:
```
def constrain_state(parameter_state: ParameterState):
    return constrain(parameter_state.params, parameter_state.bijectors)
```

Reply:
Could replace this:
```python
def loss(params):
        params = trainable_params(params, trainables)
        params = constrain(params, bijectors)
        return objective(params)
```

with something like?
```python
evaluate(objective, parameter_state)
```

Reply:
I guess we need to decide, withing GPJax, do we pass around params as a dictionary, or `ParameterState`.

Reply:
Absolutely. Doing things via `ParameterState` is clean and user friendly. It is easy to see the transformations and which parameters are currently trainable. Working with 3 unpacked dictionaries might cause more pain points if you have more than a one model on the go (having parameter state for each one would keep things tidy). The training abstractions would look cleaner. On the other hand working with 3 unpacked dictionaries is more direct with regards to the operations being applied.

Reply:
I wonder whether this is the point at which the package should split - GPJax works directly with the dictionaries and an abstracted package, such as TuneGP, works with the `ParameterState` object to give a more user-friendly interface.